### PR TITLE
arduino: use buildFHSUserEnv to support compilation of boards

### LINKED ISSUE #73965: Arduino: binaries installed through Boards Manager fail to run
**Describe the bug**
Binaries installed through the Boards Manager in Arduino IDE
fail to run due to missing shared libraries.

A solution to this is to `patchelf` them manually after installation.

**To Reproduce**
1. Tools > Boards > Boards Manager
2. install Arduino SAM Boards
3. Verify/Compile

**Expected behavior**
Code is compiled using the installed binaries

**Metadata**
Please run `nix run nixpkgs.nix-info -c nix-info -m` and paste the result.

 - system: `"x86_64-linux"`
 - host os: `Linux 4.19.80, NixOS, 19.03.173660.3f92c2124a5 (Koi)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.2`
 - channels(root): `"nixos-19.03.173660.3f92c2124a5, nixos-unstable-20.03pre201791.c1966522d7d"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
# a list of nixos modules affected by the problem
module:
```

### PR DESCRIPTION
**Pull request by Stunkymonkey**
###### Motivation for this change
fixes #73965
currently it enables compilation of other board. I tested it with ESP32.
@veprbl any comments?

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/embedded/arduino/arduino-core/default.nix [hunk 1]
```diff
@@ -1,87 +1,10 @@
...
-, pango
-, gdk-pixbuf
-, gtk2
-, libpng12
-, expat
-, freetype
-, cairo
-, udev
-}:
+{ lib, newScope, fetchFromGitHub, withGui ? false, withTeensyduino ? false, gtk3 ? null }:
```
Comment (SuperSandro2000) [inline pkgs/development/embedded/arduino/arduino-core/default.nix:1]:
```suggestion
{ lib, newScope, fetchFromGitHub, withGui ? false, withTeensyduino ? false, gtk3 }:
```

File: pkgs/development/embedded/arduino/arduino-core/core.nix [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+      patchelf --debug \
+          --set-interpreter $(cat $NIX_CC/nix-support/dynamic-linker) \
+          --set-rpath "${teensy_libpath}" \
+          $out/share/arduino/hardware/tools/teensy{,_ports,_reboot,_restart,_serialmon}
+    ''}
+  '';
+
+  meta = with lib; {
+    description = "Open-source electronics prototyping platform";
+    homepage = "http://arduino.cc/";
```
Comment (SuperSandro2000) [inline pkgs/development/embedded/arduino/arduino-core/core.nix:243]:
```suggestion
    homepage = "https://www.arduino.cc/";
```

File: pkgs/development/embedded/arduino/arduino-core/core.nix [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+    # avrdude_bin is linked against libtinfo.so.5
+    mkdir $out/lib/
+    ln -s ${lib.makeLibraryPath [ ncurses5 ]}/libtinfo.so.5 $out/lib/libtinfo.so.5
+
+    ${lib.optionalString withTeensyduino ''
+      # Patch the Teensy loader binary
+      patchelf --debug \
+          --set-interpreter $(cat $NIX_CC/nix-support/dynamic-linker) \
+          --set-rpath "${teensy_libpath}" \
+          $out/share/arduino/hardware/tools/teensy{,_ports,_reboot,_restart,_serialmon}
```
Comment (SuperSandro2000) [inline pkgs/development/embedded/arduino/arduino-core/core.nix:237]:
```suggestion
        --set-interpreter $(cat $NIX_CC/nix-support/dynamic-linker) \
        --set-rpath "${teensy_libpath}" \
        $out/share/arduino/hardware/tools/teensy{,_ports,_reboot,_restart,_serialmon}
```

File: pkgs/development/embedded/arduino/arduino-core/core.nix [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+  '';
+
+  # So we don't accidentally mess with firmware files
+  dontStrip = true;
+  dontPatchELF = true;
+
+  preFixup = ''
+    for file in $(find $out -type f \( -perm /0111 -o -name \*.so\* \) ); do
+        patchelf --set-interpreter "$(cat $NIX_CC/nix-support/dynamic-linker)" "$file" || true
+        patchelf --set-rpath ${rpath}:$out/lib $file || true
```
Comment (SuperSandro2000) [inline pkgs/development/embedded/arduino/arduino-core/core.nix:214]:
```suggestion
      patchelf --set-interpreter "$(cat $NIX_CC/nix-support/dynamic-linker)" "$file" || true
      patchelf --set-rpath ${rpath}:$out/lib $file || true
```

File: pkgs/development/embedded/arduino/arduino-core/core.nix [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+      mv arduino_dist/arduino-${version}/lib/{arduino-core,pde}.jar $out/share/arduino/lib/
+      # Delete the directory now that the jars are copied out
+      rm -r arduino_dist
+      # Extract and patch the Teensyduino installer
+      cp ${teensyduino_src} ./TeensyduinoInstall.${teensy_architecture}
+      chmod +w ./TeensyduinoInstall.${teensy_architecture}
+      upx -d ./TeensyduinoInstall.${teensy_architecture}
+      patchelf --set-interpreter $(cat $NIX_CC/nix-support/dynamic-linker) \
+          --set-rpath "${teensy_libpath}" \
+          ./TeensyduinoInstall.${teensy_architecture}
```
Comment (SuperSandro2000) [inline pkgs/development/embedded/arduino/arduino-core/core.nix:196]:
```suggestion
        --set-rpath "${teensy_libpath}" \
        ./TeensyduinoInstall.${teensy_architecture}
```

File: pkgs/development/embedded/arduino/arduino-core/core.nix [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+    cd ..
+  '';
+
+  # This will be patched into `arduino` wrapper script
+  # Java loads gtk dynamically, so we need to provide it using LD_LIBRARY_PATH
+  dynamicLibraryPath = lib.makeLibraryPath [ gtk3 ];
+  javaPath = lib.makeBinPath [ jdk ];
+
+  # Everything else will be patched into rpath
+  rpath = (lib.makeLibraryPath [ zlib libusb-compat-0_1 libusb1 readline ncurses5 stdenv.cc.cc ]);
```
Comment (SuperSandro2000) [inline pkgs/development/embedded/arduino/arduino-core/core.nix:155]:
```suggestion
  rpath = lib.makeLibraryPath [ zlib libusb-compat-0_1 libusb1 readline ncurses5 stdenv.cc.cc ];
```

File: pkgs/development/embedded/arduino/arduino-core/core.nix [hunk 7]
```diff
@@ -0,0 +1,248 @@
...
+  # Used because teensyduino requires jars be a specific size
+  arduino_dist_src = fetchurl {
+    url = "https://downloads.arduino.cc/arduino-${version}-${teensy_architecture}.tar.xz";
+    sha256 =
+      {
+        linux64 = "sha256-VK+Skl2xjqPWYEEKt1CCLwBZRxoyRfYQ3/60Byen9po=";
+        linux32 = "sha256-fjqV4avddmWAdFqMuUNUcDguxv3SI45m5QHFiWP8EKE=";
+        linuxarm = "sha256-Br8vUN7njI7VCH+ZvUh44l8LcgW+61+Q0x2AiXxIhTM=";
+        linuxaarch64 = "sha256-bOizBUUuyINg0/EqEatBq9lECT97JXxKbesCGyCA3YQ=";
+      }.${teensy_architecture} or (throw "No arduino binaries for ${teensy_architecture}");
```
Comment (SuperSandro2000) [inline pkgs/development/embedded/arduino/arduino-core/core.nix:107]:
```suggestion
    sha256 = {
      linux64 = "sha256-VK+Skl2xjqPWYEEKt1CCLwBZRxoyRfYQ3/60Byen9po=";
      linux32 = "sha256-fjqV4avddmWAdFqMuUNUcDguxv3SI45m5QHFiWP8EKE=";
      linuxarm = "sha256-Br8vUN7njI7VCH+ZvUh44l8LcgW+61+Q0x2AiXxIhTM=";
      linuxaarch64 = "sha256-bOizBUUuyINg0/EqEatBq9lECT97JXxKbesCGyCA3YQ=";
    }.${teensy_architecture} or (throw "No arduino binaries for ${teensy_architecture}");
```

File: pkgs/development/embedded/arduino/arduino-core/core.nix [hunk 8]
```diff
@@ -0,0 +1,248 @@
...
+    zlib
+  ];
+  teensy_architecture =
+    if stdenv.hostPlatform.isx86_32 then "linux32"
+    else if stdenv.hostPlatform.isx86_64 then "linux64"
+    else if stdenv.hostPlatform.isAarch64 then "linuxaarch64"
+    else if stdenv.hostPlatform.isAarch32 then "linuxarm"
+    else throw "${stdenv.hostPlatform.system} is not supported in teensy";
+
+
```
Comment (SuperSandro2000) [inline pkgs/development/embedded/arduino/arduino-core/core.nix:83]:
```suggestion
    else throw "${stdenv.hostPlatform.system} is not supported in teensy";

```

File: pkgs/development/embedded/arduino/arduino-core/core.nix [hunk 9]
```diff
@@ -0,0 +1,248 @@
...
+, expat
+, freetype
+, cairo
+, udev
+, pname
+, version
+, src
+}:
+
+assert withGui -> gtk3 != null && wrapGAppsHook != null;
```
Comment (SuperSandro2000) [inline pkgs/development/embedded/arduino/arduino-core/core.nix:37]:
```suggestion
```

File: pkgs/development/embedded/arduino/arduino-core/core.nix [hunk 10]
```diff
@@ -0,0 +1,248 @@
...
+, jdk
+, ant
+, libusb-compat-0_1
+, libusb1
+, unzip
+, zlib
+, ncurses
+, readline
+, withGui ? false
+, gtk3 ? null
```
Comment (SuperSandro2000) [inline pkgs/development/embedded/arduino/arduino-core/core.nix:14]:
```suggestion
, gtk3
```

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

File: pkgs/development/embedded/arduino/arduino-core/chrootenv.nix [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+    platforms = platforms.linux;
+    maintainers = with maintainers; [ stunkymonkey ];
+  };
+
+  extraInstallCommands = ''
+    find ${src}
+    echo "###"
+    find $out
+    mkdir -p $out/share/applications
+    # cp ${src}/share/applications/arduino.desktop $out/share/applications/arduino.desktop
```
Comment (veprbl) [inline pkgs/development/embedded/arduino/arduino-core/chrootenv.nix:36]: It might help to use `makeDesktopItem`, if the original one doesn't work.

Comment (Stunkymonkey) [reply pkgs/development/embedded/arduino/arduino-core/chrootenv.nix:36]: i am unable to find it, so yes good idea

Comment (Stunkymonkey) [reply pkgs/development/embedded/arduino/arduino-core/chrootenv.nix]: no idea, why i did not find it yesterday. I just saw it today and fixed the desktop-file with its icon. can you please verify, that it works?

**Comment by Stunkymonkey** [comment]
any idea on how to fix `octave-6.3.0-arduino-0.7.0`?

**Comment by Stunkymonkey** [comment]
is the last commit a good workaround?

File: pkgs/development/embedded/arduino/arduino-core/chrootenv.nix (hunk 11 above)
Comment (veprbl) [reply pkgs/development/embedded/arduino/arduino-core/chrootenv.nix]: I don't have a Linux setup to test this right now. I think that as long as it works for you, and the diff looks good we just merge this PR.

**Comment by Stunkymonkey** [comment]
@SuperSandro2000 is the last commit a common practice? or how to solve this?

File: pkgs/top-level/all-packages.nix [hunk 12]
```diff
@@ -1255,7 +1255,9 @@ with pkgs;
 
   arduino-cli = callPackage ../development/embedded/arduino/arduino-cli { };
 
-  arduino-core = callPackage ../development/embedded/arduino/arduino-core { };
+  arduinoPackages = dontRecurseIntoAttrs (callPackage ../development/embedded/arduino/arduino-core { });
```
Comment (veprbl) [inline pkgs/top-level/all-packages.nix:1258]: Do you even need this extra scope at this point?

Comment (Stunkymonkey) [reply pkgs/top-level/all-packages.nix:1258]: i do not know for sure. I copied it from the `platformio`. Should i remove it?

File: pkgs/development/embedded/arduino/arduino-core/chrootenv.nix [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+    ]);
+
+in
+buildFHSUserEnv {
+  name = "${pname}-${version}";
+
+  targetPkgs = arduino-pkgs;
+  multiPkgs = null;
+
+  meta = with lib; {
```
Comment (SuperSandro2000) [inline pkgs/development/embedded/arduino/arduino-core/chrootenv.nix:23]: meta should be last.

**Comment by SuperSandro2000** [comment]
> @SuperSandro2000 is the last commit a common practice? or how to solve this?

I am not sure, buildUseFHSUserEnv should only be used if there is no easier way to achieve the goal.

**Comment by Stunkymonkey** [comment]
buildUseFHSUserEnv is uesed because [`nix-environment`](https://github.com/nix-community/nix-environments/blob/master/envs/arduino/shell.nix) and [`platformio`](https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/embedded/platformio/chrootenv.nix) do it the same way.

File: pkgs/top-level/all-packages.nix (hunk 12 above)
Comment (veprbl) [reply pkgs/top-level/all-packages.nix:1258]: Yes, that should reduce the diff of this PR.

Comment (veprbl) [reply pkgs/top-level/all-packages.nix:1258]: Still there?

Comment (Stunkymonkey) [reply pkgs/top-level/all-packages.nix:1258]: no removed

Comment (veprbl) [reply pkgs/top-level/all-packages.nix:1258]: Can see it on github and this comment is not marked with "outdated"

Comment (Stunkymonkey) [reply pkgs/top-level/all-packages.nix:1258]: :open_mouth:  in the "Conversations" it does show up, but in "Files changed"/"Commits" it does not show up.

Comment (Stunkymonkey) [reply pkgs/top-level/all-packages.nix:1258]: i force pushed all again. no difference. No idea what is going on with GitHub here...

Comment (veprbl) [reply pkgs/top-level/all-packages.nix:1258]: I'm not sure what is the confusion. I do:
```
# git clone git@github.com:Stunkymonkey/nixpkgs.git -b arduino-buildFHSUserEnv
# grep arduinoPackages nixpkgs/pkgs/top-level/all-packages.nix
  arduinoPackages = callPackage ../development/embedded/arduino/arduino-core { };
  arduino-core = arduinoPackages.arduino-chrootenv;
  arduino-unwrapped = arduinoPackages.arduino-unwrapped;
```

Comment (Stunkymonkey) [reply pkgs/top-level/all-packages.nix:1258]: i am talking about `dontRecurseIntoAttrs`, which is no longer present. What is your suggested change?

Comment (Stunkymonkey) [reply pkgs/top-level/all-packages.nix:1258]: something like this?

Comment (veprbl) [reply pkgs/top-level/all-packages.nix:1258]: I was talking about dropping the use of the `newScope`

Comment (Stunkymonkey) [reply pkgs/top-level/all-packages.nix:1258]: sorry for the misunderstanding.
It does not compile without. Either I am doing it wrong, or it is needed.

**Comment by SuperSandro2000** [comment]
I don't really have the time right now to review such a big diff.

**Comment by veprbl** [comment]
I've made some changes according to my reviews, the diff should be manageable now. @Stunkymonkey, can you test that this still works for you?

**Comment by Stunkymonkey** [comment]
some changes broke it.
it is no longer called `arduino` but `arduino-core` instead.
The desktop icon therefore no longer works, because it expects the resulting binary to be called `arduino` and not `arduino-core-1.8.16`
When starting the binary `arduino-core-1.8.16`; some weird stuff happens:
```
/nix/store/p3lvcdrmy7z59fv88nk032l37acqx3sc-arduino-core-1.8.16/bin/arduino-core-1.8.16
git: symbol lookup error: /usr/lib/libc.so.6: undefined symbol: _dl_catch_error_ptr, version GLIBC_PRIVATE
hostname: symbol lookup error: /usr/lib/libc.so.6: undefined symbol: _dl_catch_error_ptr, version GLIBC_PRIVATE
sudo: symbol lookup error: /usr/lib/libc.so.6: undefined symbol: _dl_catch_error_ptr, version GLIBC_PRIVATE
/nix/store/wv35g5lff84rray15zlzarcqi9fxzz84-bash-4.4-p23/bin/bash: symbol lookup error: /usr/lib/libc.so.6: undefined symbol: _dl_catch_error_ptr, version GLIBC_PRIVATE
bash: /key-bindings.bash: No such file or directory
/nix/store/wv35g5lff84rray15zlzarcqi9fxzz84-bash-4.4-p23/bin/bash: symbol lookup error: /usr/lib/libc.so.6: undefined symbol: _dl_catch_error_ptr, version GLIBC_PRIVATE
bash: /completion.bash: No such file or directory
direnv: symbol lookup error: /usr/lib/libc.so.6: undefined symbol: _dl_catch_error_ptr, version GLIBC_PRIVATE
/nix/store/wv35g5lff84rray15zlzarcqi9fxzz84-bash-4.4-p23/bin/bash: symbol lookup error: /usr/lib/libc.so.6: undefined symbol: _dl_catch_error_ptr, version GLIBC_PRIVATE
```
it also spawns a new bash-shell.

File: pkgs/development/embedded/arduino/arduino-core/chrootenv.nix [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+      substituteInPlace $out/share/applications/arduino.desktop \
+        --replace '<BINARY_LOCATION>' "$out/bin/arduino" \
+        --replace '<ICON_NAME>' "$out/share/arduino/icons/128x128/apps/arduino.png"
+      # icon file
+      mkdir -p $out/share/arduino
+      cp -r ${arduino-core-unwrapped.src}/build/shared/icons $out/share/arduino
+    ''}
+  '';
+
+  passthru.runScript = "arduino";
```
Comment (veprbl) [inline pkgs/development/embedded/arduino/arduino-core/chrootenv.nix:31]: I think should not have added passthru here
```suggestion
  runScript = "arduino";
```

Comment (Stunkymonkey) [reply pkgs/development/embedded/arduino/arduino-core/chrootenv.nix:31]: still results in `arduino-core`

**Comment by Stunkymonkey** [comment]
even by removing the `passthru` the naming is still wrong.
Am I getting it right, that the amount of changes (in my PR) is to high to review?

Maybe i can split my commits, such that the changes I made are better visible.

**Comment by veprbl** [comment]
It's basically an UX issue with github/git diff display, but it's resolved now. Splitting into separate commits was another option, but there was also issue that the `newScope` is unnecessary and has a potential to slow down eval. I think fixing the current state of this PR would be a better option.

File: pkgs/development/embedded/arduino/arduino-core/chrootenv.nix [hunk 15]
```diff
@@ -0,0 +1,34 @@
+{ lib, buildFHSUserEnv, arduino-core-unwrapped, withGui ? false, withTeensyduino ? false }:
+
+buildFHSUserEnv {
+  name = "${arduino-core-unwrapped.pname}-${arduino-core-unwrapped.version}";
+
+  targetPkgs =
+    pkgs: (with pkgs; [
+      ncurses
+      arduino-core-unwrapped
```
Comment (Stunkymonkey) [inline pkgs/development/embedded/arduino/arduino-core/chrootenv.nix:9]: currently there is a problem with the package `arduino` from "all-packages.nix":
`arduino = arduino-core.override { withGui = true; };`
it used to result into the package `arduino-1.8.16` and now results into `arduino-core-1.8.16` which could be weird to debug.

Comment (Stunkymonkey) [reply pkgs/development/embedded/arduino/arduino-core/chrootenv.nix:9]: same problem with the parameter: `withTeensyduino`

**Comment by Stunkymonkey** [comment]
@veprbl what do you think about my changes? do they make sense to you? I will test this more on Sunday, but the first compilation with ESP32 works.

**Comment by veprbl** [comment]
Looks great. This will need to be rebased. Please let us know when this is ready to merge.

**Comment by Stunkymonkey** [comment]
should i squash all commits into a single one? i do not really like the diffs in between each commit.

**Comment by veprbl** [comment]
Yes, please squash and rebase on top of the "master".

**Comment by Stunkymonkey** [comment]
@veprbl from my point of view it is ready to be merged.

**Comment by Stunkymonkey** [comment]
if you want, I can separate this into three commits for better understanding.

**Comment by veprbl** [comment]
Result of `nixpkgs-review pr 144772` [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>6 packages built:</summary>
<br>- arduino
<br>- arduino-core
<br>- arduino-core-unwrapped
<br>- ino
<br>- octavePackages.arduino
<br>- teensyduino
</details>

**Review by veprbl — APPROVED** [review]

**Comment by Stunkymonkey** [comment]
ping @SuperSandro2000

**Pull request merged**